File: ios/Classes/Video.m
//
//  Video.m
//  opentok_flutter
//
//  Created by Ayon Das on 20/06/22.
//

#import <Foundation/Foundation.h>
#import <Flutter/Flutter.h>
#import "Video.h"

@implementation OpenTokVideoContainer

- (instancetype)initWithFrame:(CGRect)frame {
  self = [super initWithFrame:frame];
  [self setAutoresizesSubviews:true];
  [self setAutoresizingMask:(UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight)];
  return self;
}

- (void)layoutSubviews {
  [super layoutSubviews];
}

@end

@implementation OpentokVideoPlatformView
OpenTokVideoContainer *videoContainer;

- (instancetype)initWithFrame:(CGRect)frame {
  videoContainer = [[OpenTokVideoContainer alloc] initWithFrame:frame];
  self = [super init];
  return self;
}

- (nonnull UIView *)view {
  return videoContainer;
}

- (void) addSubscriberView : (UIView *) view {
  [videoContainer insertSubview:view atIndex:0];
  [view setTranslatesAutoresizingMaskIntoConstraints: NO];
  NSLayoutConstraint *horizontal = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeLeading
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:videoContainer
                                    attribute:NSLayoutAttributeLeading
                                    multiplier:1.0
                                    constant:0];
  NSLayoutConstraint *vertical = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeTop
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:videoContainer
                                    attribute:NSLayoutAttributeTop
                                    multiplier:1.0
                                    constant:0];
  NSLayoutConstraint *width = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeWidth
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:videoContainer
                                    attribute:NSLayoutAttributeWidth
                                    multiplier:1.0
                                    constant:0];
  NSLayoutConstraint *height = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeHeight
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:videoContainer
                                    attribute:NSLayoutAttributeHeight
                                    multiplier:1.0
                                    constant:0];
  [videoContainer addConstraints:[NSArray arrayWithObjects:horizontal, vertical, width, height, nil]];
}

- (void) addPublisherView : (UIView *) view {
  [videoContainer addSubview:view];
  [view setTranslatesAutoresizingMaskIntoConstraints: NO];
  NSLayoutConstraint *horizontal = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeRight
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:videoContainer
                                    attribute:NSLayoutAttributeRight
                                    multiplier:1.0
                                    constant:-20];
  NSLayoutConstraint *vertical = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeBottom
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:videoContainer
                                    attribute:NSLayoutAttributeBottom
                                    multiplier:1.0
                                    constant:-70];
  NSLayoutConstraint *width = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeWidth
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:nil
                                    attribute:NSLayoutAttributeNotAnAttribute
                                    multiplier:1.0
                                    constant:120];
  NSLayoutConstraint *height = [NSLayoutConstraint
                                    constraintWithItem:view
                                    attribute:NSLayoutAttributeHeight
                                    relatedBy:NSLayoutRelationEqual
                                    toItem:nil
                                    attribute:NSLayoutAttributeNotAnAttribute
                                    multiplier:1.0
                                    constant:160];
  [videoContainer addConstraints:[NSArray arrayWithObjects:horizontal, vertical, width, height, nil]];
}

@end

@implementation OpentokVideoFactory
@synthesize subscriberView = _subscriberView;
@synthesize publisherView = _publisherView;

- (instancetype)initWithBinaryMessenger:(id<FlutterBinaryMessenger>)binaryMessenger {
  self = [super init];
  if (self) {
    _binaryMessenger = binaryMessenger;
  }
  return self;
}

- (OpentokVideoPlatformView *)getViewInstance:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args messenger:(id<FlutterBinaryMessenger>)binaryMessenger {
  if (_view == nil) {
    _view = [[OpentokVideoPlatformView alloc] initWithFrame:frame];
    if (_subscriberView != nil) {
      [_view addSubscriberView:_subscriberView];
    }
    if (_publisherView != nil) {
      [_view addPublisherView:_publisherView];
    }
  }
  return _view;
}

- (nonnull NSObject<FlutterPlatformView> *)createWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args { 
  return [self getViewInstance:frame viewIdentifier:viewId arguments:args messenger:_binaryMessenger];
}

@end

File: ios/Classes/Video.swift
//
//  Video.swift
//  opentok_flutter
//
//  Created by Ayon Das on 13/06/22.
//

import Foundation

class OpentokVideoFactory: NSObject, FlutterPlatformViewFactory {
    static var view: OpentokVideoPlatformView?

    static var viewToAddSub: UIView?
    static var viewToAddPub: UIView?

    static func getViewInstance(
        frame: CGRect,
        viewId: Int64,
        args: Any?,
        messenger: FlutterBinaryMessenger?
    ) -> OpentokVideoPlatformView{
        if(view == nil) {
            view = OpentokVideoPlatformView()
            if viewToAddSub != nil {
                view?.addSubscriberView(viewToAddSub!)
            }
            if viewToAddPub != nil {
                view?.addPublisherView(viewToAddPub!)
            }
        }

        return view!
    }

    private var messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return OpentokVideoFactory.getViewInstance(
            frame: frame,
            viewId: viewId,
            args: args,
            messenger: messenger)
    }
}

class OpentokVideoPlatformView: NSObject, FlutterPlatformView {
    private let videoContainer: OpenTokVideoContainer

    override init() {
        videoContainer = OpenTokVideoContainer()
        super.init()
    }

    public func addSubscriberView(_ view: UIView) {
        videoContainer.addSubscriberView(view)
    }

    public func addPublisherView(_ view: UIView) {
        videoContainer.addPublisherView(view)
    }

    func view() -> UIView {
        return videoContainer
    }
}

final class OpenTokVideoContainer: UIView {
    private let subscriberContainer = UIView()
    private let publisherContainer = UIView()

    init() {
        super.init(frame: .zero)
        addSubview(subscriberContainer)
        addSubview(publisherContainer)
    }


    public func addSubscriberView(_ view: UIView) {
        subscriberContainer.addSubview(view)
    }

    public func addPublisherView(_ view: UIView) {
        publisherContainer.addSubview(view)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        let width = frame.width
        let height = frame.height

        let videoWidth = width / 2
        subscriberContainer.frame = CGRect(x: 0, y: 0, width: videoWidth, height: height)
        publisherContainer.frame = CGRect(x: videoWidth, y: 0, width: videoWidth, height: height)
    }
}
